src/webgpu/api/operation/rendering/3d_texture_slices.spec.ts

```diff
@@ -0,0 +1,363 @@
+    options: {
+      format?: GPUTextureFormat;
+      attachmentCount?: number;
+    } = {}
+  ): GPURenderPipeline {
+    const { format = 'rgba8unorm', attachmentCount = 1 } = options;
+
+    let locations = '';
+    let outputs = '';
+    const targets: GPUColorTargetState[] | null = [];
+    for (let i = 0; i < attachmentCount; i++) {
+      locations = locations + `@location(${i}) color${i} : vec4f, \n`;
+      outputs = outputs + `output.color${i} = vec4f(0.0, 1.0, 0.0, 1.0);\n`;
+      targets.push({ format } as GPUColorTargetState);
+    }
+
+    return this.device.createRenderPipeline({
+      layout: 'auto',
+      vertex: {
+        module: this.device.createShaderModule({
+          code: `
+          @vertex fn main(
+            @builtin(vertex_index) VertexIndex : u32
+            ) -> @builtin(position) vec4<f32> {
+              var pos : array<vec2<f32>, 3> = array<vec2<f32>, 3>(
+                  vec2<f32>(-1.0, 1.0),
+                  vec2<f32>(1.0, -1.0),
+                  vec2<f32>(-1.0, -1.0));
+              return vec4<f32>(pos[VertexIndex], 0.0, 1.0);
+            }
+            `,
+        }),
+        entryPoint: 'main',
+      },
+      fragment: {
+        module: this.device.createShaderModule({
+          code: `
+            struct Output {
+              ${locations}
+            }
+            @fragment fn main() -> Output {
+              var output : Output;
+              ${outputs}
+              return output;
+            }
+            `,
+        }),
+        entryPoint: 'main',
+        targets,
+      },
+      primitive: { topology: 'triangle-list' },
+    });
+  }
+
+  getBufferSizeAndOffset(
+    attachmentWidth: number,
+    attachmentHeight: number,
+    attachmentCount: number
+  ): { bufferSize: number; bufferOffset: number } {
+    const bufferSize =
+      (attachmentCount * attachmentHeight - 1) * kBytesPerRowAlignment + attachmentWidth * 4;
+    const bufferOffset = attachmentCount > 1 ? attachmentHeight * kBytesPerRowAlignment : 0;
+    return { bufferSize, bufferOffset };
+  }
+
+  checkAttachmentResult(
+    attachmentWidth: number,
+    attachmentHeight: number,
+    attachmentCount: number,
+    buffer: GPUBuffer
+  ) {
+    const { bufferSize, bufferOffset } = this.getBufferSizeAndOffset(
+      attachmentWidth,
+      attachmentHeight,
+      attachmentCount
+    );
+    const expectedData = new Uint8Array(bufferSize);
+    for (let i = 0; i < attachmentCount; i++) {
+      for (let j = 0; j < attachmentHeight; j++) {
+        for (let k = 0; k < attachmentWidth; k++) {
+          expectedData[i * bufferOffset + j * 256 + k * 4] = k < j ? 0x00 : 0xff;
+          expectedData[i * bufferOffset + j * 256 + k * 4 + 1] = k < j ? 0xff : 0x00;
+          expectedData[i * bufferOffset + j * 256 + k * 4 + 2] = 0x00;
+          expectedData[i * bufferOffset + j * 256 + k * 4 + 3] = 0xff;
+        }
+      }
+    }
+
+    this.expectGPUBufferValuesEqual(buffer, expectedData);
+  }
```

Comment:
Consider using `expectTexelViewComparisonIsOkInTexture` with `TexelView.fromTexelsAsColors` rather than having to construct the buffer layout for the copy.

(But if it's not straightforward, don't spend too much time on it; this is fine too.)

Reply:
I think we still need to construct the pixel color array for fromTexelsAsColors, I‘d like to keep it as is.

Reply:
OK.